Support HIPSYCL_TARGETS in cmake (rebased)

psath opened the pull request.

psath committed:
Support HIPSYCL_TARGETS in cmake

psath committed:
Fix: Cmake: Only check for HIPSYCL_PLATFORM if HIPSYCL_TARGETS is unset

illuhad commented on cmake/hipsycl-config.cmake.in:
```diff
@@ -48,40 +48,66 @@ if(HIPSYCL_ROCM_BACKEND_AVAILABLE)
+  set (HIPSYCL_TARGETS_CANONICAL "${HIPSYCL_TARGETS}")
+  string(REGEX REPLACE "cpu|host|hipcpu|openmp|omp" "omp" HIPSYCL_TARGETS_CANONICAL "${HIPSYCL_TARGETS_CANONICAL}")
+  string(REGEX REPLACE "cuda|nvidia" "cuda" HIPSYCL_TARGETS_CANONICAL "${HIPSYCL_TARGETS_CANONICAL}")
+  string(REGEX REPLACE "rocm|amd|hip|hcc" "hip" HIPSYCL_TARGETS_CANONICAL "${HIPSYCL_TARGETS_CANONICAL}")
```
As I've said previously, it's not clear to me why we need this here - `syclcc --hipsycl-targets` does not support the old synonyms, and I think we should align what cmake supports with what `syclcc` supports. That way, we can deprecate and remove the synonyms in the future altogether.

psath replied:
Thanks for the reminder! Had other fires to put out this week, but I'll try to push a commit stripping this canonicalization out before the end of the day

psath replied:
Fixed in 2fa9df2

psath committed:
fixup! Remove CMake HIPSYCL_TARGETS canonicalization, since the synonyms are deprecated

psath committed:
fixup! fixup! Remove CMake HIPSYCL_TARGETS canonicalization, since the synonyms are deprecated

psath committed:
Fix: Correct ROCm path and CXX host compiler setting to work with non-canonicalized HIPSYCL_TARGETS

illuhad commented:
Okay, I think this is almost ready to go, I've only noticed some minor issues with the documentation:
* `HIPSYCL_PLATFORM_STRING` should mention `hip`, not `rocm`, as backend for consistency with `HIPSYCL_TARGETS` which only accepts `hip`.
* Maybe mention in the argument description that `HIPSYCL_PLATFORM` and `HIPSYCL_GPU_ARCH` are deprecated?
* When running `cmake` without arguments, it asks the user to specify `HIPSYCL_PLATFORM` - it should ask to specify `HIPSYCL_TARGETS` as `HIPSYCL_PLATFORM` is deprecated.

psath committed:
Clarify deprecation of HIPSYCL_PLATFORM and HIPSYCL_GPU_ARCH
Replace remaining HIPSYCL_TARGETS reference to rocm with the correct hip backend

psath commented:
> Okay, I think this is almost ready to go, I've only noticed some minor issues with the documentation:
> 
> * `HIPSYCL_PLATFORM_STRING` should mention `hip`, not `rocm`, as backend for consistency with `HIPSYCL_TARGETS` which only accepts `hip`.
> * Maybe mention in the argument description that `HIPSYCL_PLATFORM` and `HIPSYCL_GPU_ARCH` are deprecated?
> * When running `cmake` without arguments, it asks the user to specify `HIPSYCL_PLATFORM` - it should ask to specify `HIPSYCL_TARGETS` as `HIPSYCL_PLATFORM` is deprecated.

04acfcb should handle all the above. I added a `HIPSYCL_PLATFORMS` canonicalization of `omp` --> `cpu` (while `HIPSYCL_PLATFORMS` remains supported) as `targets` only accepts `omp`, so `HIPSYCL_PLATFORM_STRING` needs to refer to it, not `cpu`. (`--hipsycl-platforms` already accepted `omp` but cmake wouldn't appropriately detect it to set `HIPSYCL_PLATFORMS_CANONICAL`)

illuhad commented:
> 04acfcb should handle all the above. I added a PLATFORMS canonicalization of omp --> cpu (while PLATFORMS remains supported) as targets only accepts omp, so HIPSYCL_PLATFORM_STRING needs to refer to it, not cpu. (--hipsycl-platforms already accepted omp but cmake wouldn't appropriately detect it to set PLATFORMS_CANONICAL)

Thanks, that makes sense.

@hipsycl-ci-bot test this as well please.

EDIT: Run again, my dear bot!

illuhad commented:
Okay, hipSYCL-ci-bot fails because this branch does not yet have PR #383 merged which is required for compilation with clang 11 which, among other things, is tested. So this is nothing to be concerned about.

psath committed:
Merge remote-tracking branch 'origin/develop' into develop

illuhad commented:
Wonderful, thank you very much for this great contribution!
